string-ot-v1

geyang opened the pull request.

symboli-artoria17 committed:
update transform_del_move() and makeOpMov()

geyang commented on workspace/package.json:
```diff
@@ -22,6 +22,7 @@
   "author": "Jiaming Guo, Cha Chen, Ge Yang",
   "license": "UNLICENSED",
   "devDependencies": {
+    "babel-cli": "^6.23.0",
```
babel cli should not be inside devDependency. @jiamingguo

geyang replied:
@jiaming please address this comment!

geyang commented on workspace/src/string-transform.js:
```diff
@@ -38,13 +38,139 @@ function transform_ins_ins(op0, op1) {
+  //   [_____]     [--------]
+  //  0 1 2 3 4 5 6 7 8 9 10 11 12 13 ...
+  //    op1           op0(del)
+
```
These graphs are awesome! don't get carried away though~

geyang commented on workspace/src/string-transform.js:
```diff
@@ -38,13 +38,139 @@ function transform_ins_ins(op0, op1) {
+
+function transform_del_mov(op0, op1) {
+  // case 0: destination of Mov located before deletion area
+  if((op1.pos+op1.length<=op0.pos && op1.des <= op0.pos)||(op1.pos >= op0.pos+op0.length && (op1.des >= op0.pos+op0.length || op1.des <= op0.pos))){
```
These lines are way too long. Limit lines to less than 80 characters

geyang commented on workspace/src/string-transform.js:
```diff
@@ -38,13 +38,139 @@ function transform_ins_ins(op0, op1) {
+  //   {____[===}------]
+  //  0 1 2 3 4 5 6 7 8 9 10 11 12 13 ...
+  //    op1       op0(del)
+  if (op1.pos<=op0.pos && op1.pos+op1.length<=op0.pos+op0.length&& op1.pos+op1.length >= op0.pos && op1.des>=op0.pos+op0.length){
```
all of these logics can be combined. 

1. Use single layer of `if` `else if`
2. combine cases
3. maybe assign some of the conditions to a variable.

geyang commented on workspace/src/utility.js:
```diff
@@ -14,11 +14,11 @@ export function makeOpDel (pos, length) {
     type : "mov",
     pos : pos,
-    value : value,
+    value : length,
```
this shouldn't be `value: length` right? @jam-world @jiamingguo

symboli-artoria17 replied:
I changed it to length. Because I found length is easier to work with.

geyang replied:
The problem is not length. The problem is that you are still using `value` as the key.

geyang replied:
I think it should be value: value.

@jam-world in `es6`, you do `{value}` and it is the same as `value: value`.

symboli-artoria17 replied:
oh boy. Your are right. I'm correcting it to length:length

geyang replied:
@jiamingguo did you see my comment on `ES6` syntax?

geyang reviewed (commented):
- logic need to be combined 
- `babel-cli` need to be removed from package.json.

symboli-artoria17 committed:
1. Updated String-ops 'mov': Previous version seems not correct
2. Combined logic of transform_del_mov and added new cases
3. Created block comment to wrap the graphs in transform_del_mov so it could be folded (looks better)
4. utility.js: makeOpMov {length:length} instead of {value:length}

geyang commented on workspace/src/string-ops.js:
```diff
@@ -7,6 +7,16 @@ export function del(s, i, l) {
 
 export function mov(s, i, l, d) {
-    return s.slice(0, i) + s.slice(i+l, d) + s.slice(d, l);
+  if (d>=i+l){
```
Format this code with IntelliJ/webstorm to comply with standard. (Add spaces)

geyang commented on workspace/src/string-transform.js:
```diff
@@ -38,13 +38,207 @@ function transform_ins_ins(op0, op1) {
+     //    op1        op0(del)
+
+     */
+    if (op1.des >= op1.pos + l1 && op1.des <= op0.pos + l0) {
```
Take a look at the transformIndex function. Can these logic be packaged and reused that function?

geyang commented on workspace/src/string-transform.js:
```diff
@@ -60,18 +254,17 @@ export function transform(op0, op1) {
       return transform_del_del(op0, op1);
     } else if (op1.type === "mov") {
-      // todo: del then mov
-      return nil;
```
@jam-world nil is not a JavaScript token ha~

symboli-artoria17 committed:
reformat the code

symboli-artoria17 committed:
removed babel-cli from package.json

geyang commented on workspace/src/string-transform.js:
```diff
@@ -53,25 +247,24 @@ export function transform(op0, op1) {
   if (op0.type === "ins") {
```
you could potentially use a dictionary to realize this function. @jiamingguo @jam-world 

Use 
```
transforms = {'ins,del': transform_ins_del, ...}
```

geyang replied:
@jiamingguo can you see if you can try this?

geyang reviewed (commented):
looks good to me!

symboli-artoria17 committed:
added all test cases for transform_del_mov (All tests pass)

jiamingguo commented:
![20170310231055](https://cloud.githubusercontent.com/assets/15992269/23889177/13957bf6-0862-11e7-96a2-4a5957b5d5ef.png)

jiamingguo commented:
![testcases](https://cloud.githubusercontent.com/assets/15992269/23889216/45e45ab4-0862-11e7-9bd0-0324c624b95d.PNG)

symboli-artoria17 reviewed (commented):
![testcases](https://cloud.githubusercontent.com/assets/15992269/23889260/7a3aa692-0862-11e7-8442-6437d12661ab.PNG)

symboli-artoria17 committed:
added function: transform_ins_mov

symboli-artoria17 committed:
added figures for transform_ins_mov case 3

geyang commented on workspace/src/string-transform.js:
```diff
@@ -20,31 +20,335 @@ export function transformCursor(cursor, op) {
+
+function transform_del_mov(op0, op1) {
+  var l1 = op1.length;
+  var l0 = op0.length;
```
should use `const` for all of these `var`s

geyang commented on workspace/src/string-transform.js:
```diff
@@ -20,31 +20,335 @@ export function transformCursor(cursor, op) {
+  var l1 = op1.length;
+  var des = op1.des;
+  var pos0 = op0.pos;
+  var pos1 = op1.pos;
```
should use `const` for all of these `var`s

symboli-artoria17 replied:
Just did it. See next commit

geyang commented on workspace/src/utility.js:
```diff
@@ -1,24 +1,24 @@
-export function makeOpIns (pos, value) {
+export function makeOpIns(pos, value) {
```
we can rename these to `Ins`, `Del`, `Mov`.

symboli-artoria17 replied:
Is it necessary to change? That function name is good though.

geyang replied:
If it is only used in test we don't have to change it. It does feel a bit redundant.

geyang committed:
change test script from test to test-watch, use  to run test once

geyang committed:
update README with test and test-watch

episodeyang commented:
![screenshot 2017-03-17 00 25 20](https://cloud.githubusercontent.com/assets/630490/24029988/44fa500c-0aa8-11e7-9158-3a15a2b4c7a7.png）

ops test have not passed, so the transform code is likely to be incorrect as well. @jiamingguo

geyang commented:
![screenshot 2017-03-17 00 25 20](https://cloud.githubusercontent.com/assets/630490/24029988/44fa500c-0aa8-11e7-9158-3a15a2b4c7a7.png)

ops test have not passed, so the transform code is likely to be incorrect as well. @jiamingguo

geyang committed:
ignore dist in jest

geyang committed:
add default value to del

geyang commented:
@jiamingguo You tests are okay! those failed tests came from the `dist` folder. I have updated the npm test script to avoid those folders.

geyang commented on workspace/src/string-ops.js:
```diff
@@ -2,7 +2,7 @@ export function ins(s, i, v) {
 }
 
-export function del(s, i, l) {
+export function del(s, i, l = 1) {
```
@jam-world Otherwise the slice i + l would return `NaN`.

symboli-artoria17 replied:
I will add a new test case for this.

geyang replied:
that is just the default for `l`.

symboli-artoria17 committed:
Added del-string-at-beginning2. Test the case that is not using default l=1.

geyang committed:
README with tentative usage pattern

geyang committed:
Merge branch 'ot-test' of https://github.com/escherpad/scala_spark_ot into ot-test

geyang commented:
Looks pretty good to me, we can probably merge this. 

Milestone `string-ot-v1` complete? @jiamingguo

symboli-artoria17 committed:
Added new test case

geyang committed:
change operation creator name

geyang committed:
Merge branch 'ot-test' of https://github.com/escherpad/scala_spark_ot into ot-test

geyang committed:
remove semicolon

geyang committed:
move all string type files into ot-string

geyang committed:
add utils for OpSet

geyang committed:
add OpSet test

geyang committed:
use OpSet instead of array

geyang committed:
remove unused dependency

geyang commented:
this one is ready to be merged. @jiamingguo
